vecmath/include/vecmath/mat.h

```diff
@@ -786,6 +794,109 @@ namespace vm {
         }
         return result;
     }
+
+    /**
+     * Finds an LUP decomposition of matrix a.
+     *
+     * Give A, finds P,L,U satisfying PA=LU where P is a permutation matrix,
+     * where L is lower-triangular with the diagonal elements set to 1,
+     * U is upper-triangular.
+     *
+     * The permutation matrix is returned in a compressed form where each element of the vector represents a row of
+     * the permutation matrix, and a value of `i` means the `i`th column of that row is set to 1.
+     *
+     * From "LUP-Decomposition", Introduction to Algorithms by Cormen et. al., 2nd. ed. p752.
+     *
+     * @tparam T the component type
+     * @tparam S the number of components
+     * @param a the matrix to decompose
+     * @return {true, L and U packed into a single matrix, compressed permutation matrix}
+     *         or {false, unspecified, unspecified} if a decomposition doesn't exist.
+     */
+    template <typename T, size_t S>
+    std::tuple<bool, mat<T,S,S>, vec<size_t,S>> lupDecomposition(mat<T,S,S> a) {
+        using std::swap;
+
+        vec<size_t,S> pi;
+        for (size_t i=0; i<S; ++i) {
+            pi[i] = i;
+        }
+        for (size_t k=0; k<S; ++k) {
+            T p(0);
+            size_t kPrime = 0;
+            for (size_t i=k; i<S; ++i) {
+                if (vm::abs(a[k][i]) > p) {
+                    p = vm::abs(a[k][i]);
+                    kPrime = i;
+                }
+            }
+            if (p == 0) {
+                return {false, {}, {}};
```

Comment:
Wasn't there a problem with this syntax with some compiler(s)? I'd prefer `std::make_tuple(false, vm::mat4x4(), vm::vec3())`

Reply:
I think it was just for std::pair that you need std::make_pair instead of {a,b}. std::make_tuple sounds fine to me though.